### PR TITLE
Merchant new item#13

### LINKED ISSUE #13: User Story 54, Merchant adds an item
```
As a merchant
When I visit my items page
And I click on the link to add a new item
My URI route should be "/dashboard/items/new"
I see a form where I can add new information about an item, including:
- the name of the item, which cannot be blank
- a description for the item, which cannot be blank
- a thumbnail image URL string, which CAN be left blank
- a price which must be greater than $0.00
- my current inventory count of this item which is 0 or greater

When I submit valid information and save the form
I am taken back to my items page
I see a flash message indicating my new item is saved
I see the new item on the page, and it is enabled and available for sale
If I left the image field blank, I see a placeholder image for the thumbnail
```

### Checklist:

- [ ] base code tests are written
- [ ] base code is written to pass tests
- [ ] edge case ("sad path") tests written (if required)
- [ ] edge case code written (if required)
- [ ] code is reviewed
- [ ] branch is merged
- [ ] all tests still pass

child of #18

depends on #17

### PR DESCRIPTION
**Pull request by Bradniedt**
-Adds tests for merchant: 
1- seeing the add new item link
2- creating an item with valid information
3- creating an item without an image link
4- creating an item with a negative inventory (sad path)
-Implemented functionality for card 13
-All tests passing 
-99.95% coverage

closes #13

### CODE REVIEW

File: db/migrate/20181215211701_create_items.rb [hunk 1]
```diff
@@ -6,7 +6,7 @@ def change
       t.integer :inventory
       t.string :description
       t.integer :current_price
-      t.boolean :enabled
+      t.boolean :enabled, default: true
```
Comment (danbriechle) [inline db/migrate/20181215211701_create_items.rb]: Little bitty one here but is this going to bite us later on? I guess we can cross that bridge when we get there :)

Comment (Bradniedt) [reply db/migrate/20181215211701_create_items.rb]: @danbriechle I don't believe it will, because we'll have the ability to disable the item. Also here's the snippet from the user story ```When I submit valid information and save the form
I am taken back to my items page
I see a flash message indicating my new item is saved
I see the new item on the page, and it is enabled and available for sale``` 
So it looks like items are supposed to be enabled by default per this story.

Comment (danbriechle) [reply db/migrate/20181215211701_create_items.rb]: sounds good to me

File: app/models/item.rb [hunk 2]
```diff
@@ -21,4 +24,16 @@ def self.five_popular(var)
           .order("total_orders #{var}")
           .limit(5)
   end
+
+  def self.enabled_items
+    Item.all.where(enabled: true)
```
Comment (asmolentzov) [inline app/models/item.rb]: Small, but do you need the ‘Item.all’ here? Should just be able to do ‘where’...

File: db/migrate/20181215211701_create_items.rb (hunk 1 above)
Comment (saleratus) [reply db/migrate/20181215211701_create_items.rb]: I think this confusion came from something we heard in class. Ian said that in a usual production environment, items are usually created an NOT enabled, so that the merchant or site planner can get it properly created and set up before going live. He also specified that we are not going to do this in our Little Shop project.

File: app/models/item.rb [hunk 3]
```diff
@@ -1,12 +1,15 @@
 class Item < ApplicationRecord
-  validates_presence_of :name, :image_link, :inventory, :description,
-                        :current_price, :enabled, :user_id
-
+  validates_presence_of :name, :image_link, :description,
+                        :current_price, :user_id
+  validates :enabled, inclusion: {in: [true, false]}
+  validates :inventory, inclusion: {in: (1..1000000)}
```
Comment (asmolentzov) [inline app/models/item.rb]: Nice!!

File: app/views/items/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,20 @@
+<%= form_for @item, :url => dashboard_items_new_path(@item) do |f| %>
```
Comment (asmolentzov) [inline app/views/items/new.html.erb]: I think the standard path for this would be dashboard_items_path, with a method of post (which should be default here)

File: db/migrate/20181215211701_create_items.rb (hunk 1 above)
Comment (Bradniedt) [reply db/migrate/20181215211701_create_items.rb]: @saleratus - oh yeah that's where we heard that - you're totally right!

File: config/routes.rb [hunk 5]
```diff
@@ -15,6 +15,8 @@
   get '/dashboard/items', to: "items#index"
   put '/dashboard/item', to: "items#update"
   delete '/dashboard/item', to: "items#destroy"
+  get '/dashboard/items/new', to: "items#new"
+  post '/dashboard/items/new', to: "items#create"
```
Comment (asmolentzov) [inline config/routes.rb]: See above comment - I don’t think we want the /new on this

File: spec/models/item_spec.rb [hunk 6]
```diff
@@ -79,5 +92,12 @@
         expect(Item.five_popular('asc')).to_not include(item_3)
       end
     end
+    describe '#set_default_image' do
+      it 'sets a default image path before saving if one is not present' do
+        user = create(:merchant)
+        item = user.items.create(name: 'apple1', image_link: '', inventory: 3, description: 'apple one', current_price: 200, enabled: true)
+        expect(item.image_link).to eq('https://picsum.photos/200/300?image=0')
```
Comment (asmolentzov) [inline spec/models/item_spec.rb]: Might want to add a test here to make sure it doesn’t add the default image to an item that does already have an image link

File: app/views/items/index.html.erb [hunk 7]
```diff
@@ -10,6 +10,10 @@
   <% end %>
 </ul>
 
+<% if current_merchant? %>
+  <%= link_to 'Add New Item', '/dashboard/items/new' %>
```
Comment (asmolentzov) [inline app/views/items/index.html.erb]: Should prob use path helper here

File: app/models/item.rb [hunk 8]
```diff
@@ -1,12 +1,15 @@
...
   belongs_to :user
 
   has_many :order_items
   has_many :orders, through: :order_items
 
+  before_validation :set_default_image
```
Comment (asmolentzov) [inline app/models/item.rb]: This is cool!!

File: spec/features/merchant_new_item_spec.rb [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+      expect(current_path).to eq('/dashboard/items/new')
+
+      fill_in :item_name, with: 'item_1'
+      fill_in :item_description, with: 'this is a great item_1'
+      fill_in :item_image_link, with: 'https://picsum.photos/200'
+      fill_in :item_inventory, with: 1
+      click_on('Create Item')
+
+      expect(current_path).to eq('/dashboard/items/new')
+      expect(page).to have_content("Current price can't be blank")
```
Comment (asmolentzov) [inline spec/features/merchant_new_item_spec.rb]: Might want to add another block under here to check for the rest of the error messages. Can do them all in one go.

File: spec/features/merchant_new_item_spec.rb [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+      expect(current_path).to eq('/dashboard/items/new')
+
+      fill_in :item_name, with: 'item_1'
+      fill_in :item_description, with: 'this is a great item_1'
+      fill_in :item_current_price, with: 3000
+      fill_in :item_inventory, with: -1
+      click_on('Create Item')
+
+      expect(current_path).to eq('/dashboard/items/new')
+      expect(page).to have_content('Inventory is not included in the list')
```
Comment (asmolentzov) [inline spec/features/merchant_new_item_spec.rb]: Depends on what the user story says, but do you think it would be appropriate to add an additional flash message to the error message here, something like "Your item was not created"?

File: app/models/item.rb (hunk 2 above)
Comment (Bradniedt) [reply app/models/item.rb]: You're totally right - good catch! for some reason I was getting stuck thinking I needed to call the where on all. Changing now!

File: app/views/items/new.html.erb (hunk 4 above)
Comment (Bradniedt) [reply app/views/items/new.html.erb]: Hmm - the user story called for the /new:
```When I visit my items page
And I click on the link to add a new item
My URI route should be "/dashboard/items/new"```
Should I change it? Or keep it as is?

File: spec/models/item_spec.rb (hunk 6 above)
Comment (Bradniedt) [reply spec/models/item_spec.rb]: I add a different image in the first test - should I add an `expect(page).to_not have` for the default image on that test?

File: app/views/items/index.html.erb (hunk 7 above)
Comment (Bradniedt) [reply app/views/items/index.html.erb]: You're totally right - I had forgotten to change the path once I made the route. Changed. Thanks!

File: app/models/item.rb (hunk 8 above)
Comment (Bradniedt) [reply app/models/item.rb]: Gracias!! =]

File: spec/features/merchant_new_item_spec.rb (hunk 9 above)
Comment (Bradniedt) [reply spec/features/merchant_new_item_spec.rb]: Facepalm! Good eye - I added all to that test - quick fix. Thanks!

File: app/views/items/new.html.erb (hunk 4 above)
Comment (Bradniedt) [reply app/views/items/new.html.erb]: Changed!

Comment (Bradniedt) [reply app/views/items/new.html.erb]: Yep - the order in the routes file was messing with the path helper prefix.  Moved it and it is all good now!

File: config/routes.rb (hunk 5 above)
Comment (Bradniedt) [reply config/routes.rb]: Resolved per above.

File: spec/models/item_spec.rb (hunk 6 above)
Comment (Bradniedt) [reply spec/models/item_spec.rb]: I added a separate test for creating a good item without an image, and also added an expectation to the first test to expect to not see the default image. =]
